src/core/crypto.c

```diff
@@ -1719,10 +1722,68 @@ QuicCryptoCustomCertValidationComplete(
             "Custom cert validation failed.");
         QuicConnTransportError(
             QuicCryptoGetConnection(Crypto),
-            QUIC_ERROR_CRYPTO_ERROR(0xFF & CXPLAT_TLS_ALERT_CODE_BAD_CERTIFICATE));
+            QUIC_ERROR_CRYPTO_ERROR(0xFF & CXPLAT_TLS_ALERT_CODE_BAD_CERTIFICATE)); //
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicCryptoCustomTicketValidationComplete(
+    _In_ QUIC_CRYPTO* Crypto,
+    _In_ BOOLEAN Result
+    )
+{
+    //
+    // Finalizes server side resumption ticket validation in case
+    // server app decide to validate it asynchronously
+    // Need to set TicketValidationPending = FALSE to drain packet and continue handshake
+    //
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    if (!Crypto->TicketValidationPending || Crypto->TicketValidationRejecting) {
+        Status = QUIC_STATUS_INVALID_STATE;
+        goto Error;
+    }
+
+    if (Result) {
+        //
+        // Just call completion.
+        // Outgoing buffer was already prepared during previous initial packet processing
+        //
+        Crypto->TicketValidationPending = FALSE;
+        QuicCryptoProcessDataComplete(Crypto, Crypto->PendingValidationBufferLength);
+    } else {
+        //
+        // Need to rollback status before processing client's initial packet, because outgoing buffer and
+        // related status was speculatively set for successfull case as above
+        //
+
+        //
+        // TicketValidationRejecting is for intuitive naming purpose in QuicConnRecvResumptionTicket
+        // TicketValidationPending = FALSE will be set at QuicConnRecvResumptionTicket
+        //
+        Crypto->TicketValidationRejecting = TRUE;
+
+        Crypto->TlsState.ReadKey = QUIC_PACKET_KEY_INITIAL;
+        Crypto->TlsState.WriteKey = QUIC_PACKET_KEY_INITIAL;
+        Crypto->TlsState.BufferOffsetHandshake = 0;
+        Crypto->TlsState.BufferOffset1Rtt = 0;
+        for (size_t i = QUIC_PACKET_KEY_0_RTT; i < QUIC_PACKET_KEY_COUNT; ++i) {
+            QuicPacketKeyFree(Crypto->TlsState.ReadKeys[i]);
+            Crypto->TlsState.ReadKeys[i] = NULL;
+            QuicPacketKeyFree(Crypto->TlsState.WriteKeys[i]);
+            Crypto->TlsState.WriteKeys[i] = NULL;
+        }
+        Crypto->RecvBuffer.ExternalBufferReference = FALSE;
+        QUIC_CONNECTION* Connection = QuicCryptoGetConnection(Crypto);
+        Status = QuicCryptoInitializeTls(Crypto, Connection->Configuration->SecurityConfig, Connection->HandshakeTP);
```

Comment:
If this fails for some reason, we'll need to fatal this connection, via `QuicConnFatalError`.